Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcCopyClient.cs

```diff
@@ -67,9 +129,27 @@ private GrpcCopyClient(string host, int grpcPort, bool useCompression)
         /// <param name="useCompression">Whether or not GZip is enabled for copies.</param>
         public static GrpcCopyClient Create(string host, int grpcPort, bool useCompression = false)
         {
-            // TODO: Add caching of GrpcCopyClient objects
-            // TODO: Add case where _clientPool has exceeded some maximum count
-            return new GrpcCopyClient(host, grpcPort, useCompression);
+            if (_clientDict.TryGetValue((host, grpcPort, useCompression), out GrpcCopyClient existingClient))
+            {
+                Interlocked.Increment(ref existingClient._uses);
+                existingClient._lastUseTime = DateTime.UtcNow;
+                return existingClient;
+            }
+            else if (_clientDict.Count > _maxClientCount)

```

Comment:
>ntDict.Count > _maxClie [](start = 26, length = 23)

Curious why you didn't do this check inside the GetOrAdd lambda, now there is a race where many threads can get past this check and blow the nr of Clients beyand the maxClient so now it is kind of like a soft limmit.

Reply:
`GetOrAdd`'s factory argument doesn't affect whether the factory's result will be added. We could guard with a lock, but it's okay if this is only a soft limit.

---
In reply to: [289194267](https://github.com/microsoft/BuildXL/pull/391#discussion_r289194267) [](ancestors = 289194267)